### PR TITLE
wrapped UDTs;

### PR DESCRIPTION
**Pull request by TobyAM**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -294,17 +319,21 @@ void CoffeeShop::printNumBaristas()
...
+        JUCE_LEAK_DETECTOR(Invoice)
     };
 
 
-    Invoice createInvoice(std::string clientName, float dueDate, std::string workType, float workTime);
-    bool checkOverdue(Invoice invoice);
-    float checkBalance(Invoice invoice);
+    Invoice* createInvoice(const std::string clientName, const float dueDate, const std::string workType, const float workTime);
+    bool checkOverdue(const Invoice* invoice);
+    float checkBalance(const Invoice* invoice);
```
Comment (matkatmusic) [inline main.cpp]: Pass by reference, not by pointer.
- [x] revise

File: main.cpp [hunk 2]
```diff
@@ -294,17 +319,21 @@ void CoffeeShop::printNumBaristas()
         float workTime, totalBalance, dueDate;
         bool overdue;
 
-        void download(std::string format);
-        void markAsPaid(Invoice& invoiceA);
-        void duplicate(Invoice& invoiceA);
+        void download(const std::string format);
+        void markAsPaid(Invoice* invoiceA);
```
Comment (matkatmusic) [inline main.cpp]: Pass by reference, not by pointer.
- [x] revise

File: main.cpp [hunk 3]
```diff
@@ -102,16 +102,18 @@ struct CoffeeShop
...
     };
 
-    Coffee brewCoffee( std::string customerName, Coffee coffeeType, int size, std::string brewType, bool cream, bool sugar);
-    bool grindCoffee(Coffee coffeeType, int coarseness, Customer customerA);
-    void renameCustomer(Customer& customerA, std::string newName); // changed to a reference to fix, IDK
-    void pourCoffee(Coffee coffee, int size);
+    Coffee brewCoffee(const std::string customerName, const Coffee& coffeeType, const int size, const std::string brewType, const bool cream, const bool sugar);
+    bool grindCoffee(const Coffee& coffeeType, const int coarseness, const Customer* customerA);
+    void renameCustomer(Customer* customerA, const std::string newName); // changed to a reference to fix, IDK
+    void pourCoffee(const Coffee& coffee, const int size);
```
Comment (matkatmusic) [inline main.cpp]: Pass by reference, not by pointer.
- [x] revise

File: main.cpp [hunk 4]
```diff
@@ -561,9 +639,11 @@ struct Studio
     ScooterRental scooterA, scooterB;
     InvoiceManager invoices;
 
-    void orderRun(std::string runner, InvoiceManager::Invoice invoice, ScooterRental scooter, int latCoord, int longCoord, std::string order);
-    void prepareCoffee(std::string runner, std::string clientName, CoffeeShop::Coffee coffeeType, int size, std::string brewType, bool cream, bool sugar);
+    void orderRun(const std::string runner, const InvoiceManager::Invoice* invoice, const ScooterRental& scooter, const int latCoord, const int longCoord, const std::string order);
+    void prepareCoffee(const std::string runner, const std::string clientName, const CoffeeShop::Coffee& coffeeType, const int size, const std::string brewType, const bool cream, const bool sugar);
```
Comment (matkatmusic) [inline main.cpp]: Pass by reference, not by pointer.
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp [hunk 5]
```diff
@@ -357,26 +397,26 @@ void InvoiceManager::Invoice::markAsPaid(Invoice& invoiceA)
...
 
-void InvoiceManager::Invoice::duplicate(Invoice& invoiceA)
+void InvoiceManager::Invoice::duplicate(const Invoice& invoiceA)
 {
-    Invoice invoiceB = invoiceA;    // I would not really do this
-    ++invoiceB.invoiceNumber;
-    std::cout << "Duplicated invoice " << invoiceA.invoiceNumber << " as " << invoiceB.invoiceNumber << std::endl;
+    Invoice* invoiceB = new Invoice(invoiceA);
+    ++invoiceB->invoiceNumber;
+    std::cout << "Duplicated invoice " << invoiceA.invoiceNumber << " as " << invoiceB->invoiceNumber << std::endl;
```
Comment (matkatmusic) [inline main.cpp]: No.
What happens when you hit the closing curly brace of this function? 
you allocated this object on the heap.
Did you de-allocate it? 
- [x] revise

File: main.cpp [hunk 6]
```diff
@@ -357,26 +397,26 @@ void InvoiceManager::Invoice::markAsPaid(Invoice& invoiceA)
...
-    newInvoice.dueDate = date;
-    newInvoice.workType = type;
-    newInvoice.workTime = time;
+    newInvoice->invoiceNumber = numInvoices;
+    newInvoice->dueDate = date;
+    newInvoice->workType = type;
+    newInvoice->workTime = time;
 
     return newInvoice;
 }
```
Comment (matkatmusic) [inline main.cpp]: Don't allocate this object on the heap. 
You are not de-allocating it anywhere.
The point of this assignment is to learn how to manually manage heap-allocated objects in a safe manner using Wrapper classes. 
- [x] revert to returning the invoice as a regular stack object.

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp (hunk 5 above)
Comment (TobyAM) [reply main.cpp]: I spent some time trying to make this function valid, but ultimately it seems to be a nonsensical prototype that should be deleted at this point.

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**